pkg/controller/externalnode/controller.go

```diff
@@ -0,0 +1,384 @@
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeAdd(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode ADD event", "ExternalNode", klog.KObj(en))
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeUpdate(oldObj interface{}, newObj interface{}) {
+	en := newObj.(*v1alpha1.ExternalNode)
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode UPDATE event", "ExternalNode", klog.KObj(en))
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeDelete(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	key, _ := keyFunc(en)
+	c.queue.Add(key)
+	klog.InfoS("Enqueued ExternalNode DELETE event", "ExternalNode", klog.KObj(en))
+}
+
+// Run will create defaultWorkers workers (goroutines) which will process the ExternalEntity events from the work queue.
+func (c *ExternalNodeController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.externalNodeListerSynced) {
+		klog.Error("Failed to wait for syncing cache for ExternalNodes")
+		return
+	}
+	if err := c.reconcileExternalNode(); err != nil {
+		klog.Errorf("Failed to reconcile ExternalNodes %v", err)
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// reconcileExternalNode reconciles all the existing ExternalNodes and cleans up the stale ExternalEntities.
+func (c *ExternalNodeController) reconcileExternalNode() error {
+	externalNodes, err := c.externalNodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	enNameSet := sets.NewString()
+	for _, en := range externalNodes {
+		enNameSet.Insert(en.Name)
+		if err = c.addExternalNode(en); err != nil {
+			return err
+		}
+	}
+	externalEntities, err := c.externalEntityLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, ee := range externalEntities {
+		if len(ee.OwnerReferences) > 0 {
+			if !enNameSet.Has(ee.OwnerReferences[0].Name) {
+				klog.InfoS("Deleting stale ExternalEntity", "ExternalEntity", klog.KObj(ee))
+				err = c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Delete(context.TODO(), ee.Name, metav1.DeleteOptions{})
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem function in
+// order to read and process a message on the work queue.
+func (c *ExternalNodeController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *ExternalNodeController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in ExternalNode work queue but got %#v", obj)
```

Comment:
ErrorS

Please change other code too

Reply:
I think this one cannot use ErrorS as there is no error generated here. But I will update all the others.